File: 1353. Maximum Number of Events That Can Be Attended/maximum_number_of_events_that_can_be_attended.cpp
// Sorting and Heap
// T.C.->O(nlogn)
// S.C.->O(n)
class Solution {
public:
    int maxEvents(vector<vector<int>>& events) {
        int n = events.size();
        sort(begin(events), end(events));
        priority_queue<int, vector<int>, greater<int>> pq; //min-heap
        int day = events[0][0];
        int i=0;
        int count=0;
        while(!pq.empty() || i<n) {
            while(i<n && events[i][0] == day) {
                pq.push(events[i][1]);
                i++;
            }

            if(!pq.empty()) {
                pq.pop(); //1 event attended on this day
                count++; //counting the result
            }
            day++;

            //skip those events whose endDay < day
            while(!pq.empty() && pq.top() < day) {
                pq.pop();
            }
        }
        return count;
    }
};

File: 1353. Maximum Number of Events That Can Be Attended/maximum_number_of_events_that_can_be_attended.py
# Sorting and Heap
# T.C.->O(nlogn)
# S.C.->O(n)
class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        events.sort()  # Sort by start day
        n = len(events)
        min_heap = []  # Acts as a min-heap for end days
        day = events[0][0]
        i = 0
        count = 0

        while min_heap or i < n:
            # Add all events starting today
            while i < n and events[i][0] == day:
                heapq.heappush(min_heap, events[i][1])
                i += 1

            # Attend one event today
            if min_heap:
                heapq.heappop(min_heap)
                count += 1

            day += 1

            # Remove events whose end day has passed
            while min_heap and min_heap[0] < day:
                heapq.heappop(min_heap)

        return count
